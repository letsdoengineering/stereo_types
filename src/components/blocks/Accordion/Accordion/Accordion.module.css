.accordion-section {
    display: flex;
    flex-direction: column;
    width: 100%;
}

.accordion {
    background-color: inherit;
    padding: 0 0 0.5rem 0;
    cursor: pointer;
    display: flex;
    align-items: center;
    border: none;
    outline: none;
    width: 100%;
}

.accordion--icon-right {
    justify-content: space-between;
}

button.accordion:focus {
    outline: 5px auto Highlight;
    outline: 5px auto -webkit-focus-ring-color;
}

.rotate {
    transform: rotate(180deg);
}

.accordion-content {
    background-color: inherit;
    overflow: hidden;
    transition: max-height 0.1s ease;
}

.icon {
    margin: 0 0.5rem 0 0.5rem;
    right: 0;
}
