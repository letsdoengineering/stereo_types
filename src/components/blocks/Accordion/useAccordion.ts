import React, { useRef, useState } from 'react'
import * as self from './useAccordion'

type UseAccordion = {
  /** indicates whether the accordion should be open or not */
  isOpen: boolean
  /** used to set isOpen state */
  setIsOpen: React.Dispatch<React.SetStateAction<boolean>>
  /** value of the scrollHeight of the div containing the accordion's content
   *  e.g. '10px' used to expand the accordion */
  scrollHeight: string
  /** used to set the scrollHeight state */
  setScrollHeight: React.Dispatch<React.SetStateAction<string>>
  /** a React ref for use with the expandable/collapsable element in the accordion */
  ref: React.RefObject<HTMLDivElement>
  /** sets isOpen state and scrollHeight */
  toggleAccordion(): void
}

/** getScrollHeight: returns a string indicating scrollHeight of an element based on whether it is active or not
 * @param {boolean} isActive - indicates whether to return 0px or ${scrollHeight}px
 * @param {number} scrollHeight - the scrollHeight of an element - e.g. as taken from an element's ref */
export const getScrollHeight = (isActive: boolean, scrollHeight: number | undefined): string => {
  return isActive ? '0px' : `${scrollHeight}px`
}

/** useAccordion: custom hook returns an object containing state for use with
 * accordion components and a toggleAccordion function
 * @returns { object } UseAccordion */
export default (): UseAccordion => {
  const [isOpen, setIsOpen] = useState(false)
  const [scrollHeight, setScrollHeight] = useState('0px')
  const ref = useRef<HTMLDivElement>(null)

  const toggleAccordion = (): void => {
    setIsOpen(!isOpen)
    setScrollHeight(self.getScrollHeight(isOpen, ref?.current?.scrollHeight))
  }

  return {
    isOpen,
    setIsOpen,
    scrollHeight,
    setScrollHeight,
    ref,
    toggleAccordion,
  }
}
