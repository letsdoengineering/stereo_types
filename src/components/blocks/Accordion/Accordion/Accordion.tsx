import React, { useEffect } from 'react'
import './Accordion.module.css'
import Icon from '../../../basics/Icon/Icon'
import Text from '../../../basics/Text/Text'
import getClassNames from '../../../../utils/get-class-names'
import * as accordionStyles from './Accordion.module.css'
import Heading from '../../../basics/Heading/Heading'
import useAccordion from '../useAccordion'

type AccordionProps = {
    /** the title of the accordion */
    title: string
    /** what size to display the accordion title */
    titleSize: ElementSize
    /** the section heading tag level between <h1>-<h6> that you wish to set */
    titleLevel?: '1' | '2' | '3' | '4' | '5' | '6'
    /** set to true if Accordion should start in Active state */
    startOpen?: boolean
    /** icon should space all the way to the right of the accordion's container */
    iconRight?: boolean
    /** children always of type ReactNode */
    children?: React.ReactNode
    /** a string to be used to form new unique IDs from whilst keeping a sense of common identity */
    baseId: string
}

/** Accordion: returns an accordion that can be clicked by a user to expand the contents passed as children */
const Accordion: React.FC<AccordionProps> = ({
    title,
    titleSize,
    titleLevel = '3',
    startOpen = false,
    iconRight = false,
    children,
    baseId,
}: AccordionProps) => {
    const { isOpen, scrollHeight, toggleAccordion, ref } = useAccordion()

    useEffect(() => {
        if (startOpen) toggleAccordion()
    }, [])

    const iconClasses = getClassNames({
        defaultClasses: [accordionStyles.icon],
        conditionalClasses: {
            [`${accordionStyles.rotate}`]: isOpen,
            [`${accordionStyles.accordionIconRight}`]: iconRight,
        },
    })

    const accordionClasses = getClassNames({
        defaultClasses: [accordionStyles.accordion],
        conditionalClasses: {
            [`${accordionStyles.accordionIconRight}`]: iconRight,
        },
    })

    const contentId = `${baseId}-content`
    const accordionId = `${baseId}-accordion`

    return (
        <>
            <div className={accordionStyles.accordionSection}>
                <Heading level={titleLevel}>
                    <button
                        type='button'
                        id={accordionId}
                        aria-expanded={isOpen}
                        aria-controls={contentId}
                        className={accordionClasses}
                        onClick={toggleAccordion}
                    >
                        <Text tagName='span' size={titleSize} weight='Bold'>
                            {title}
                        </Text>
                        <Icon
                            icon='chevron-down'
                            iconSize={titleSize}
                            iconColor='Grey'
                            className={iconClasses}
                        />
                    </button>
                </Heading>
                <div
                    id={contentId}
                    aria-labelledby={accordionId}
                    ref={ref}
                    role='region'
                    style={{ maxHeight: `${scrollHeight}` }}
                    className={accordionStyles.accordionContent}
                >
                    {children}
                </div>
            </div>
        </>
    )
}

export default Accordion
