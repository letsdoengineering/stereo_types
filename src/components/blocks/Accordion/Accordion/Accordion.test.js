import * as AccordionComponent from './Accordion'
import * as useAccordion from '../useAccordion'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { axe } from 'jest-axe'
import * as React from 'react'
const Accordion = AccordionComponent.default

describe('Accordion', () => {
    afterEach(() => {
        jest.clearAllMocks()
    })

    const AccordionProps = {
        title: 'Some Accordion',
        titleSize: 'S',
        baseId: 'accordion',
    }

    it('should adjust the scrollheight of the content container when user clicks the accordion button', () => {
        const getScrollHeightSpy = jest.spyOn(useAccordion, 'getScrollHeight')
        render(<Accordion {...AccordionProps}>Some text</Accordion>)

        userEvent.click(screen.getByText('Some Accordion'))
        expect(getScrollHeightSpy).toHaveBeenCalledTimes(1)
    })

    it('should toggle to open if startOpen is set to true', () => {
        const getScrollHeightSpy = jest.spyOn(useAccordion, 'getScrollHeight')
        render(
            <Accordion startOpen {...AccordionProps}>
                Some text
            </Accordion>
        )

        expect(getScrollHeightSpy).toHaveBeenCalledTimes(1)
    })

    it('has not changed the JSX layout of the page from previous snapshot - default', () => {
        const tree = renderer.create(<Accordion {...AccordionProps} />).toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('has not changed the JSX layout of the page from previous snapshot - iconRight = true', () => {
        const tree = renderer.create(<Accordion iconRight {...AccordionProps} />).toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders the page DOM that is accessible', async () => {
        const { container } = render(<Accordion {...AccordionProps}>Some text</Accordion>)
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})
