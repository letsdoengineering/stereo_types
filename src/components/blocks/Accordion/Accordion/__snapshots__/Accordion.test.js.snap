// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion has not changed the JSX layout of the page from previous snapshot - default 1`] = `
<div
  className="accordionSection"
>
  <h3
    className="title title3 weightNormal"
  >
    <button
      aria-controls="accordion-content"
      aria-expanded={false}
      className="accordion"
      id="accordion-accordion"
      onClick={[Function]}
      type="button"
    >
      <span
        className="sizeS weightBold textGray"
      >
        Some Accordion
      </span>
      <span
        aria-hidden="true"
        className="fas fa-chevron-down iconSizeS iconColorGrey icon"
      />
    </button>
  </h3>
  <div
    aria-labelledby="accordion-accordion"
    className="accordionContent"
    id="accordion-content"
    role="region"
    style={
      Object {
        "maxHeight": "0px",
      }
    }
  />
</div>
`;

exports[`Accordion has not changed the JSX layout of the page from previous snapshot - iconRight = true 1`] = `
<div
  className="accordionSection"
>
  <h3
    className="title title3 weightNormal"
  >
    <button
      aria-controls="accordion-content"
      aria-expanded={false}
      className="accordion accordionIconRight"
      id="accordion-accordion"
      onClick={[Function]}
      type="button"
    >
      <span
        className="sizeS weightBold textGray"
      >
        Some Accordion
      </span>
      <span
        aria-hidden="true"
        className="fas fa-chevron-down iconSizeS iconColorGrey icon accordionIconRight"
      />
    </button>
  </h3>
  <div
    aria-labelledby="accordion-accordion"
    className="accordionContent"
    id="accordion-content"
    role="region"
    style={
      Object {
        "maxHeight": "0px",
      }
    }
  />
</div>
`;
